File: 977.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */

/**
 * Complexity:
 * Space: O(n)
 * Time: O(n)
 */
const sortedSquares = function(nums) {
    let left = 0;
    let right = nums.length - 1;
    let count = nums.length - 1;
    const res = [];

    while (left <= right) {
        if (Math.abs(nums[left]) < Math.abs(nums[right])) {
            res[count] = nums[right] * nums[right];
            count--;
            right--;
        } else {
            res[count] = nums[left] * nums[left];
            left++;
            count--;
        }
    }
    return res;
};

File: 977.py
'''
Complexity:
Space: O(n)
Time: O(n)
'''
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [0] * n

        left = 0
        right = n - 1
        for i in range(n-1, -1, -1):
            if abs(nums[left]) < abs(nums[right]):
                result[i] = nums[right] * nums[right]
                right -= 1
            else:
                result[i] = nums[left] * nums[left]
                left += 1
        return result
